bentoml/deployment/sagemaker/__init__.py

```diff
@@ -343,27 +358,216 @@ def delete(self):
+                "Image": ecr_image_path,
+                "Environment": {"API_NAME": sagemaker_config.api_name},
+            },
+            "ExecutionRoleArn": execution_role_arn,
+        }
+        logger.info("Creating sagemaker model %s", model_name)
+        create_model_response = sagemaker_client.create_model(**sagemaker_model_info)
+        logger.info("AWS create model response: %s", create_model_response)
+
+        production_variants = [
+            {
+                "VariantName": generate_aws_compatible_string(
+                    deployment_spec.bento_name
+                ),
+                "ModelName": model_name,
+                "InitialInstanceCount": sagemaker_config.instance_count,
+                "InstanceType": sagemaker_config.instance_type,
+            }
+        ]
+        endpoint_config_name = create_sagemaker_endpoint_config_name(
+            deployment_spec.bento_name, deployment_spec.bento_version
+        )
+        logger.info(
+            "Creating Sagemaker endpoint %s configuration", endpoint_config_name
+        )
+        create_endpoint_config_response = sagemaker_client.create_endpoint_config(
+            EndpointConfigName=endpoint_config_name,
+            ProductionVariants=production_variants,
+        )
+        logger.info(
+            "AWS create endpoint config response: %s", create_endpoint_config_response
+        )
+
+        endpoint_name = generate_aws_compatible_string(
+            deployment_pb.namespace + '-' + deployment_spec.bento_name
+        )
+        logger.info("Creating sagemaker endpoint %s", endpoint_name)
+        create_endpoint_response = sagemaker_client.create_endpoint(
+            EndpointName=endpoint_name, EndpointConfigName=endpoint_config_name
+        )
+        logger.info("AWS create endpoint response: %s", create_endpoint_response)
+
+        res_deployment_pb = Deployment(state=DeploymentState())
         res_deployment_pb.CopyFrom(deployment_pb)
-        # res_deployment_pb.state = ...
+
+        # We are going to wait 10 mins for AWS to create sagemaker
+        start_time = time.time()
+        while (time.time() - start_time) < 600:
+            state = self.describe(res_deployment_pb, repo).state
+            if state.state == DeploymentState.PENDING:
+                time.sleep(10)
+                continue
+            else:
+                break
+        res_deployment_pb.state.CopyFrom(state)
+
         return ApplyDeploymentResponse(status=Status.OK(), deployment=res_deployment_pb)
 
-    def delete(self, deployment_pb, repo):
-        # delete deployment
+    def delete(self, deployment_pb, repo=None):
+        deployment_spec = deployment_pb.spec
+        sagemaker_config = deployment_spec.sagemaker_operator_config
+        if sagemaker_config is None:
+            raise BentoMLDeploymentException('Sagemaker configuration is missing.')
+        sagemaker_client = boto3.client('sagemaker', sagemaker_config.region)
```

Comment:
think we can prepare this client beforehand so we can reuse it for both delete and apply.

Reply:
Yeah, I feel the client is redundant as well.  I put these clients here, because each actions by operator are really separate tasks, and in a multi tenants setting, it could get messy.  So I end up putting the client a lot closer to the actions instead of some place before

Reply:
Got it. Think you are right for this.